### DISCUSSION TITLE
Change Test Case Visibility By Number of Submits

### DISCUSSION THREAD

**Discussion by esherma** [gradescope/autograder_samples#48] [Q&A] [answered]
I'd like to prevent students from being able to 'debug via autograder' by simply submitting and re-submitting their code until they get 100%. I understand that one can limit the visibility of a subset of test cases so that their output and scoring only become visible after the project deadline/grade publication/etc. I'm wondering if, beyond this, it is possible to restrict visibility on test cases based on the submission count. For instance, it would be useful to have test cases either "only visible after the deadline" as well as "visible before the deadline but only on the first (X) submit(s) of the calendar day".

It does not seem, based on the documentation, that this feature is built in to the gradescope-utils functionality for reading the results.json file. If this assessment is correct, I am curious if a hand-crafted workaround is feasible to implement on my end?

Thanks in advance for your help!

**Comment by ibrahima** [comment, marked as answer]
Hi @esherma, thanks for your question! We don't have anything built into the gradescope-utils library, but you can hand-craft a solution by reading the [submission_metadata.json](https://gradescope-autograders.readthedocs.io/en/latest/submission_metadata/) file in your autograder. For something like this, since it seems like the exact mechanism for determining visibility may be fairly specialized to your use case (especially the "calendar day" part), it may not be something that makes sense for gradescope-utils's decorators.

One approach might be to sort of "post-process" the results generated by your main autograder in a second script, and adjust the visibility field according to the number of submissions the student has made.

You could also consider writing your own Python decorator based on the [default `@visibility` decorator](https://github.com/gradescope/gradescope-utils/blob/master/gradescope_utils/autograder_utils/decorators.py#L32..L51) that can do some logic based on the submission metadata, and just include that with your own testing code. For that approach, as long as you can set the `func.__visibility__` field inside the `__call__` method the JSONTestRunner will pick up on it and set it for you in the JSON output.

I'm not sure which approach would be easiest to take but I think both are decent options. This is a cool idea though and I think the solution could help other people as well! If you have any followup questions about implementation please let us know and we'll do our best to help out.

**Reply by esherma** [reply]
Thanks, @ibrahima!

I'll look into these. Can't guarantee I'll be able to get this done in any sort of timely fashion but if I do, I'll be sure to submit a pull req.

Eli
